### PR TITLE
Introducing SQSPriorityQueue

### PR DESCRIPTION
**Pull request by leogrim**
@yingjieMiao @stkem @eishay @andrewconner 
Not really a Priority Queue but sort of, feel free to come up with a better name :)

### CODE REVIEW

File: kifi-backend/modules/common/app/com/keepit/common/queue/SQSPriorityQueue.scala [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+import com.keepit.common.concurrent.FutureHelpers
+import com.keepit.common.math.ProbabilityDensityBuilder
+import com.kifi.franz.{ SQSMessage, SQSQueue }
+
+import scala.concurrent.{ Future, ExecutionContext }
+import scala.concurrent.duration.FiniteDuration
+import scala.util.Random
+
+class SQSPriorityQueue[T](weights: Map[SQSQueue[T], Double]) {
```
Comment (stkem) [inline kifi-backend/modules/common/app/com/keepit/common/queue/SQSPriorityQueue.scala]: If you extend `SQSQueue` you'll get all the other retrieval methods for free (based on `nextBatchWithLock`)

**Comment by stkem** [comment]
`ProbabilisticMultiQueue`?

File: kifi-backend/modules/common/app/com/keepit/common/queue/SQSPriorityQueue.scala [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  def nextBatchWithLock(n: Int, lockTimeout: FiniteDuration)(implicit ec: ExecutionContext): Future[Seq[SQSMessage[T]]] = {
+    val offset = offsetDensity.sample(Random.nextDouble()).get
+    // not using foldLeftWhile on purpose: it doesn't return the last accumulator, which may lock tasks without returning them
+    FutureHelpers.foldLeft(sortedQueues.drop(offset))(Seq.empty[SQSMessage[T]]) {
+      case (tasks, nextQueue) =>
+        val numberOfMissingTasks = n - tasks.length
+        if (numberOfMissingTasks <= 0) Future.successful(tasks)
+        else nextQueue.nextBatchWithLock(numberOfMissingTasks, lockTimeout).map { moreTasks => tasks ++ moreTasks }
+    }
+  }
```
Comment (yingjieMiao) [inline kifi-backend/modules/common/app/com/keepit/common/queue/SQSPriorityQueue.scala]: `sortedQueue` is sorted from high weight to low weight. When fetch jobs, offset is sampled according to the weight, and it seems that the head of the queue is dropped more often than the tail. (Thus, a queue with heaviest weight will have least probability to sample from?)

Overall I am not very clear about the intend of this method (on a quick glance).

File: kifi-backend/modules/common/app/com/keepit/common/queue/SQSPriorityQueue.scala [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+class SQSPriorityQueue[T](weights: Map[SQSQueue[T], Double]) {
+  private val sortedQueues = weights.keySet.toSeq.sortBy(-weights(_))
+  private val offsetDensity = {
+    val builder = new ProbabilityDensityBuilder[Int]()
+    sortedQueues.zipWithIndex.foreach { case (queue, index) => builder.add(index, weights(queue)) }
+    builder.build()
+  }
+  def nextBatchWithLock(n: Int, lockTimeout: FiniteDuration)(implicit ec: ExecutionContext): Future[Seq[SQSMessage[T]]] = {
+    val offset = offsetDensity.sample(Random.nextDouble()).get
+    // not using foldLeftWhile on purpose: it doesn't return the last accumulator, which may lock tasks without returning them
```
Comment (yingjieMiao) [inline kifi-backend/modules/common/app/com/keepit/common/queue/SQSPriorityQueue.scala]: If the weights are not probabilities, `sample` could return `None` with a positive probability. Cause exceptions.

Comment (leogrim) [inline kifi-backend/modules/common/app/com/keepit/common/queue/SQSPriorityQueue.scala]: I think the builder normalizes them.

File: kifi-backend/modules/common/app/com/keepit/common/queue/SQSPriorityQueue.scala (hunk 2 above)
Comment (leogrim) [inline kifi-backend/modules/common/app/com/keepit/common/queue/SQSPriorityQueue.scala]: Queues are first sampled according to their weight (I may use a `SortedSet` to make that clearer than manipulating array offsets). If that queue is exhausted, move to the next most important queue and so on.

File: kifi-backend/modules/common/app/com/keepit/common/queue/SQSPriorityQueue.scala (hunk 3 above)
Comment (yingjieMiao) [inline kifi-backend/modules/common/app/com/keepit/common/queue/SQSPriorityQueue.scala]: I see.

File: kifi-backend/modules/common/app/com/keepit/common/queue/SQSPriorityQueue.scala (hunk 1 above)
Comment (leogrim) [inline kifi-backend/modules/common/app/com/keepit/common/queue/SQSPriorityQueue.scala]: It's also asking for `fromString` and `toString`

Comment (leogrim) [inline kifi-backend/modules/common/app/com/keepit/common/queue/SQSPriorityQueue.scala]: and you can't really send, it's more of a reader

**Comment by leogrim** [comment]
`ProbabilisticMultiQueue` sounds pretty good

File: kifi-backend/modules/common/app/com/keepit/common/queue/SQSPriorityQueue.scala (hunk 2 above)
Comment (yingjieMiao) [inline kifi-backend/modules/common/app/com/keepit/common/queue/SQSPriorityQueue.scala]: `sortedQueues.drop(offset)`. The first queue is included only when offset = 0. The tail is almost always included. 

Did I miss anything?

Comment (leogrim) [inline kifi-backend/modules/common/app/com/keepit/common/queue/SQSPriorityQueue.scala]: But you only get to the tail if the previous queues haven't filled the bucket first.
Let's say we have two queues with (90%, 10%).
10% of the time we'll hit the second queue and grab whatever we can from it.
90% of the time we'll hit the first queue, and only if it's exhausted go to the second queue.

Assuming queues are always full, we get the expected (90%, 10%) distribution. Only if the most important queue is empty does the less important one benefit from a probability "spillover". Basically the actual retrieval distribution depends on the underlying queue load distribution but more important queues do prevail at full load.

**Comment by stkem** [comment]
lgtm

**Pull request merged**